internal/util/streamingutil/service/resolver/resolver_with_discoverer.go

```diff
@@ -0,0 +1,192 @@
+package resolver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/util/streamingutil/service/discoverer"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/lifetime"
+	"github.com/milvus-io/milvus/pkg/util/syncutil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+var _ Resolver = (*resolverWithDiscoverer)(nil)
+
+// newResolverWithDiscoverer creates a new resolver with discoverer.
+func newResolverWithDiscoverer(scheme string, d discoverer.Discoverer, retryInterval time.Duration) *resolverWithDiscoverer {
+	r := &resolverWithDiscoverer{
+		taskNotifier:    syncutil.NewAsyncTaskNotifier[struct{}](),
+		logger:          log.With(zap.String("scheme", scheme)),
+		registerCh:      make(chan *watchBasedGRPCResolver),
+		discoverer:      d,
+		retryInterval:   retryInterval,
+		latestStateCond: syncutil.NewContextCond(&sync.Mutex{}),
+		latestState:     d.NewVersionedState(),
+	}
+	go r.doDiscover()
+	return r
+}
+
+// versionStateWithError is the versionedState with error.
+type versionStateWithError struct {
+	state VersionedState
+	err   error
+}
+
+// resolverWithDiscoverer is the resolver for bkproxy service.
+type resolverWithDiscoverer struct {
+	taskNotifier *syncutil.AsyncTaskNotifier[struct{}]
+	logger       *log.MLogger
+
+	registerCh chan *watchBasedGRPCResolver
+
+	discoverer    discoverer.Discoverer // the discoverer method for the bkproxy service
+	retryInterval time.Duration
+
+	latestStateCond *syncutil.ContextCond
+	latestState     discoverer.VersionedState
+}
+
+// GetLatestState returns the latest state of the resolver.
+func (r *resolverWithDiscoverer) GetLatestState() VersionedState {
+	r.latestStateCond.L.Lock()
+	state := r.latestState
+	r.latestStateCond.L.Unlock()
+	return state
+}
+
+// Watch watch the state change of the resolver.
+func (r *resolverWithDiscoverer) Watch(ctx context.Context, cb func(VersionedState) error) error {
+	state := r.GetLatestState()
+	if err := cb(state); err != nil {
+		return errors.Mark(err, ErrInterrupted)
+	}
+	version := state.Version
+	for {
+		if err := r.watchStateChange(ctx, version); err != nil {
+			return errors.Mark(err, ErrCanceled)
+		}
+		state := r.GetLatestState()
+		if err := cb(state); err != nil {
+			return errors.Mark(err, ErrInterrupted)
+		}
+		version = state.Version
+	}
+}
+
+// Close closes the resolver.
+func (r *resolverWithDiscoverer) Close() {
+	// Cancel underlying task and close the discovery service.
+	r.taskNotifier.Cancel()
+	r.taskNotifier.BlockUntilFinish()
+}
+
+// watchStateChange block util the state is changed.
+func (r *resolverWithDiscoverer) watchStateChange(ctx context.Context, version typeutil.Version) error {
+	r.latestStateCond.L.Lock()
+	for version.EQ(r.latestState.Version) {
+		if err := r.latestStateCond.Wait(ctx); err != nil {
+			return err
```

Comment:
unlock r.latestStateCond.L before return error

Reply:
ContextCond is `Condvar` semantic.
when `Wait` API return error, the Mutex in ContextCond is keep `Unlocked`, so no Unlock operation should be applied.